src/rust/streaming-silent-payments/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright 2024 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![no_std]
+
+extern crate alloc;
+
+mod hash;
+
+pub use bitcoin;
+use bitcoin::hashes::Hash;
+use bitcoin::secp256k1::{self, PublicKey, Scalar, Secp256k1, SecretKey, XOnlyPublicKey};
+
+use alloc::vec::Vec;
+
+const DLEQ_PROOF_SIZE: usize = 33 + 64;
+
+pub struct Output {
+    pub pubkey: XOnlyPublicKey,
+    pub dleq_proof: [u8; DLEQ_PROOF_SIZE],
+}
+
+pub enum Network {
+    Btc,
+    Tbtc,
+}
+
+impl Network {
+    fn sp_hrp(&self) -> &str {
+        match self {
+            Network::Btc => "sp",
+            Network::Tbtc => "tsp",
+        }
+    }
+}
+
+pub struct SilentPayment {
+    secp: Secp256k1<secp256k1::All>,
+    network: Network,
+    smallest_outpoint: Option<bitcoin::OutPoint>,
+    a_sum: Option<SecretKey>,
+    // Done streaming inputs?
+    inputs_done: bool,
+    // We only allow one silent payment output for now. This tracks whether we've seen it.
+    output_created: bool,
+}
+
+fn calculate_t_k(ecdh_shared_secret: &PublicKey, k: u32) -> Result<SecretKey, ()> {
+    let hash = hash::SharedSecretHash::from_ecdh_and_k(ecdh_shared_secret, k).to_byte_array();
+    SecretKey::from_slice(&hash).map_err(|_| ())
+}
+
+fn decode_address(address: &str, expected_hrp: &str) -> Result<(PublicKey, PublicKey), ()> {
+    let mut decoded_addr =
+        bech32::primitives::decode::CheckedHrpstring::new::<bech32::Bech32m>(address)
+            .map_err(|_| ())?;
+
+    let hrp = decoded_addr.hrp();
+    let hrp: &str = hrp.as_str();
+    if hrp != expected_hrp {
+        return Err(());
+    }
+    let witness_version = decoded_addr.remove_witness_version().unwrap();
+    if witness_version != bech32::Fe32::Q {
+        return Err(());
+    }
+
+    let data: Vec<u8> = decoded_addr.byte_iter().collect();
+    if data.len() != 66 {
```

Comment:
Should magic number `66` be a const?

Reply:
added `const PUBKEY_LEN: usize = 33;`, and used `2 * PUBKEYLEN` here.